yarpl/include/yarpl/flowable/EmitterFlowable.h

```diff
@@ -26,25 +26,20 @@ class EmiterBase : public virtual Refcounted {
  * This is synchronous: the emit calls are triggered within the context
  * of a request(n) call.
  */
-template<typename T>
-class EmiterSubscription : private Subscription, private Subscriber<T> {
+template <typename T>
+class EmiterSubscription : public virtual Subscription,
+                           public virtual Subscriber<T> {
   constexpr static auto kCanceled = credits::kCanceled;
   constexpr static auto kNoFlowControl = credits::kNoFlowControl;
 
  public:
   EmiterSubscription(
-      Reference <EmiterBase<T>> emiter,
-      Reference <Subscriber<T>> subscriber)
-      : emiter_(std::move(emiter)), subscriber_(std::move(subscriber)) {
-    // We expect to be heap-allocated; until this subscription finishes
-    // (is canceled; completes; error's out), hold a reference so we are
-    // not deallocated (by the subscriber).
-    Refcounted::incRef(*this);
-    subscriber_->onSubscribe(get_ref<Subscription>(this));
-  }
+      Reference<EmiterBase<T>> emiter,
+      Reference<Subscriber<T>> subscriber)
+      : emiter_(std::move(emiter)), subscriber_(std::move(subscriber)) {}
 
-  virtual ~EmiterSubscription() {
-    subscriber_.reset();
+  void init() {
+    subscriber_->onSubscribe(get_ref<Subscription>(this));
```

Comment:
Dylan, can you elaborate a little more? I don't understand your comment. You are saying that by calling incRef in the ctor leads to calling dtor at the end of the ctor? That doesn't make sense to me. Why we didn't see this in the current code where we call onSubscribe from the ctor?